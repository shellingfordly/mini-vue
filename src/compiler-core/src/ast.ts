import { CREATE_ELEMENT_VNODE } from "./runtimeHelper";

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPUND_EXPORESS,
}

export function createVNodeCall(context, tag, props, children) {
  context.helper(CREATE_ELEMENT_VNODE);

  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children,
  };
}
