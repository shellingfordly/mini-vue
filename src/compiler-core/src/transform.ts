export function transform(root, options) {
  const context = createTransfromContext(root, options);

  traverseNode(root, context);
}

function traverseNode(node, context) {
  for (let i = 0; i < context.nodeTransform.length; i++) {
    const transform = context.nodeTransform[i];
    transform(node);
  }

  traverseChildren(node, context);
}

function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      traverseNode(child, context);
    }
  }
}

function createTransfromContext(root: any, options: any) {
  return {
    root,
    nodeTransform: options.nodeTransform || [],
  };
}
