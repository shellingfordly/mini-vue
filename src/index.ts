export * from "./reactivity/src";


export * from './runtime-core'