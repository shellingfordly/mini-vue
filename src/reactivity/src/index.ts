export {
  reactive,
  readonly,
  isProxy,
  isReactive,
  isReadonly,
} from "./reactive";

export { effect, stop } from "./effect";

export { computed } from "./computed";

export { ref, isRef, unRef, proxyRefs } from "./ref";
