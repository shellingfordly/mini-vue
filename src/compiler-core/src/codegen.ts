import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelper";
import { isString } from "../../shared";

export function generate(ast) {
  const context = createGenerateContext();
  const { push } = context;

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  push("return ");
  push(`function ${functionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");

  return context.code;
}

function genNode(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPUND_EXPORESS:
      genCompundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompundExpression(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { children } = node;
  const { push } = context;

  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function genText(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { push } = context;
  push(`"${node.content}"`);
}

function genInterpolation(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { push, helper } = context;

  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genExpression(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { push } = context;
  push(`${node.content}`);
}

function genElement(
  node: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { push, helper } = context;
  const { tag, props, children } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  genNodeList(genNullable([tag, props]), context);
  genChildren(children, context);
  push(")");
}

function genChildren(children, context) {
  const { push } = context;
  const isMoreElement = !!children.find(
    (child) => child.type === NodeTypes.ELEMENT
  );
  push(", ");
  isMoreElement && push("[ ");
  children.length ? genNodeList(children, context) : push("null");
  isMoreElement && push("] ");
}

function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];

    if (isString(node)) {
      push(node);
    } else {
      let codegenNode = node;
      if (node.type === NodeTypes.ELEMENT) {
        codegenNode = node.codegenNode;
      }
      genNode(codegenNode, context);
    }

    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullable(args: any[]) {
  return args.map((arg) => arg || "null");
}

function createGenerateContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}

function genFunctionPreamble(
  ast: any,
  context: { code: string; push(source: any): void; helper(key: any): string }
) {
  const { push } = context;
  const VueBinging = "Vue";

  if (ast.helpers.length) {
    const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`;
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`
    );
    push("\n");
  }
}
