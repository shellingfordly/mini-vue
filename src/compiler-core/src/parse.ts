import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context = createParserContext(content);

  return createRoot(parseChildren(context));
}

function parseChildren(context) {
  const nodes: any[] = [];
  const s = context.source;
  let node;

  // 处理 插值
  if (s.startsWith("{{")) {
    node = parseInterpolation(context);
  } else if (s.startsWith("<")) {
    // 处理 标签
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context);
    }
  }

  if (!node) {
    node = parseText(context);
  }

  nodes.push(node);

  return nodes;
}

function parseText(context) {
  const content = parseTextData(context, context.source.length);
  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context, length) {
  const content = context.source.slice(0, length);

  advanceBy(context, length);

  return content;
}

// 解析 element
//   处理标签 <div></div>
function parseElement(context) {
  const element = parseTag(context, TagType.Start);

  parseTag(context, TagType.End);

  return element;
}

function parseTag(context, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);

  console.log("parseTag match", context, match);

  const tag = match[1];
  // 处理完成的字符串 清除掉
  // 清除 <div
  advanceBy(context, match[0].length);
  // 清除 >
  advanceBy(context, 1);

  if (type === TagType.End) {
    return;
  }

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  // {{ message }}

  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  advanceBy(context, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;

  const rawContent = context.source.slice(0, rawContentLength);
  const content = rawContent.trim();

  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context, length: number) {
  context.source = context.source.slice(length);
}

function createRoot(children) {
  return {
    children,
  };
}

function createParserContext(children) {
  return {
    source: children,
  };
}
