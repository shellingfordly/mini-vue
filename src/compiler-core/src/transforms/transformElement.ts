import { createVNodeCall, NodeTypes } from "../ast";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const vnodeTag = `"${node.tag}"`;
      const vnodeProps = node.props;
      const vnodeChildren = node.children;

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
