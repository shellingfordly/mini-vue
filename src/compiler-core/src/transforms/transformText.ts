import { NodeTypes } from "../ast";

export function transformText(node) {
  const isText = (node) =>
    node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;

  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const { children } = node;
      let currentContainer;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];

        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextChild = children[j];

            if (isText(nextChild)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPUND_EXPORESS,
                  children: [child],
                };
              }

              currentContainer.children.push(" + ");
              currentContainer.children.push(nextChild);
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    };
  }
}
