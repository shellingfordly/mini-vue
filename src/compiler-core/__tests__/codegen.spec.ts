import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");

    transform(ast);

    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");

    transform(ast, {
      nodeTransform: [transformExpression],
    });

    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div></div>");

    transform(ast, {
      nodeTransform: [transformElement],
    });

    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("element > interpolation/string", () => {
    const ast = baseParse("<div>hi, {{message}}</div>");

    transform(ast, {
      nodeTransform: [transformExpression, transformElement, transformText],
    });

    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("element > element > interpolation/string", () => {
    const ast = baseParse("<div>hi, <span>I'm {{message}}</span></div>");

    transform(ast, {
      nodeTransform: [transformExpression, transformElement, transformText],
    });

    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });
});
