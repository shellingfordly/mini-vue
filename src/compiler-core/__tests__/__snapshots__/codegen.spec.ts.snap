// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element > element > interpolation/string 1`] = `
"const { toDisplayString: _toDisplayString, createElementVnode: _createElementVnode } = Vue
return function render(_ctx, _cache){return _createElementVnode(\\"div\\", null, [ \\"hi, \\", _createElementVnode(\\"span\\", null, \\"I'm \\" + _toDisplayString(_ctx.message))] )}"
`;

exports[`codegen element > interpolation/string 1`] = `
"const { toDisplayString: _toDisplayString, createElementVnode: _createElementVnode } = Vue
return function render(_ctx, _cache){return _createElementVnode(\\"div\\", null, \\"hi, \\" + _toDisplayString(_ctx.message))}"
`;

exports[`codegen element 1`] = `
"const { createElementVnode: _createElementVnode } = Vue
return function render(_ctx, _cache){return _createElementVnode(\\"div\\", null, null)}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache){return \\"hi\\"}"`;
